walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,140 @@
+            free_time =
+                getNewFreeTime(free_slot, event.getWhen(), free_time, conflicted_slot_index);
+            conflicted_slot_index++;
+            //break if the event ends in one of the free slots
+            if (event.getWhen().end() < free_slot.end()) break;
+          }
+        }
+      }
+    }
+
+    // If any of the TimeRange in our freetime list has a
+    // duration less than the requested meeting duration then remove that slot.
+    free_time = checkDuration(free_time, meeting_duration);
+
+    return free_time;
+  }
+
+  // This method checks if an event overlaps with any of our free time slots.
+  // It returns the conflicted slot's
+  private int containsEvent(ArrayList<TimeRange> free_time, TimeRange event_timerange) {
+    for (TimeRange slot : free_time) {
+      if (slot.overlaps(event_timerange)) return free_time.indexOf(slot);
+    }
+    return -1;
+  }
+
+  /**
+   * This method free time , conflicted free time slot and current event which we are looping
+   * through and checks all conditions in which they could overlap. Then returns the free time minus
+   * overlapping part.
+   */
+  private ArrayList<TimeRange> getNewFreeTime(
+      TimeRange free_slot,
+      TimeRange event_slot,
+      ArrayList<TimeRange> free_time,
+      int conflicted_slot_index) {
+
+    int free_slot_start = free_slot.start();
+    int free_slot_end = free_slot.end();
+
+    int event_slot_start = event_slot.start();
+    int event_slot_end = event_slot.end();
+
+    // Case: Free time: |---|
+    // Event slot:    |-------|
+    if (event_slot_start < free_slot_start && event_slot_end > free_slot_end) {
+      free_time.remove(free_slot);
+    }
+
+    // Case: Free time: |--------|
+    // Event slot:     |-----|
+    // or
+    // |------|
+    // |--|
+    else if (event_slot_start <= free_slot_start && event_slot_end < free_slot_end) {
+      free_time.remove(free_slot);
+      free_time.addAll(
+          conflicted_slot_index,
+          Arrays.asList(TimeRange.fromStartEnd(event_slot_end, free_slot_end, false)));
+    }
+
+    // Case: Free time: |--------|
+    // Event slot:            |-----|
+    // or
+    // |------|
+    //     |--|
+    else if (event_slot_start > free_slot_start && event_slot_end >= free_slot_end) {
+      free_time.remove(free_slot);
+      free_time.addAll(
+          conflicted_slot_index,
+          Arrays.asList(TimeRange.fromStartEnd(free_slot_start, event_slot_start, false)));
+    }
+
+    // Case: Free time: |--------|
+    // Event slot:        |-----|
+    else if (event_slot_start > free_slot_start && event_slot_end < free_slot_end) {
+      free_time.remove(free_slot);
+      free_time.addAll(
+          conflicted_slot_index,
+          Arrays.asList(
+              TimeRange.fromStartEnd(free_slot_start, event_slot_start, false),
+              TimeRange.fromStartEnd(event_slot_end, free_slot_end, false)));
```

Comment:
Again, I think you should use a `.set(index, ...)` followed by a `.add(index + 1, ...)`.